core/tests/test_apps.management.recovery_device.py

```diff
@@ -140,6 +140,53 @@ def test_exceptions(self):
             secret, share = process_slip39(words)
             self.assertIsNone(secret)
 
+    @mock_storage
+    def test_check_word_validity(self):
+        from trezor.messages import BackupType
+        from apps.management.recovery_device.word_validity import check, OK, NOK_IDENTIFIER_MISMATCH, NOK_ALREADY_ADDED, NOK_THRESHOLD_REACHED
+
+        storage.recovery.set_in_progress(True)
+
+        # nothing is stored -> should raise
+        with self.assertRaises(RuntimeError):
+            check(0, "ocean", BackupType.Slip39_Advanced, [])
+
+        # if backup type is not set we can not do any checks
+        result = check(0, "ocean", None, [])
+        self.assertIs(result, OK)
+
+        # BIP-39 has no "on-the-fly" checks
+        result = check(0, "ocean", BackupType.Bip39, [])
+        self.assertIs(result, OK)
+
+        # let's store two shares in the storage
+        secret, share = process_slip39("trash smug adjust ambition criminal prisoner security math cover pecan response pharmacy center criminal salary elbow bracelet lunar briefing dragon")
+        self.assertIsNone(secret)
+        secret, share = process_slip39("trash smug adjust aide benefit temple round clogs devote prevent type cards clogs plastic aspect paper behavior lunar custody intimate")
+        self.assertIsNone(secret)
+
+        # different identifier
+        result = check(0, "slush", BackupType.Slip39_Advanced, [])
+        self.assertIs(result, NOK_IDENTIFIER_MISMATCH)
+
+        # same first word but still a different identifier
+        result = check(1, "slush", BackupType.Slip39_Advanced, ["trash"])
+        self.assertIs(result, NOK_IDENTIFIER_MISMATCH)
```

Comment:
there should probably be a check for 2nd word, for completeness

Reply:
I'm confused. This a check for the 2nd word

Reply:
right. i meant 3rd word .)

Reply:
The third word contains group index, group threshold and a part of the group count. We currently check only if the threshold was reached and that is tested at the end. We might extend this by checking if the group count/threshold for that group index fits with the previous one, but it seems a bit superfluous.

Reply:
ah, but now I realize that what is actually missing here is `Slip39_Basic` test

Reply:
Added. We are responding with `IdentifierMismatch` which is not 100% accurate. It probably should be `GroupsMetadataMismatch`. But maybe it's fine for now? Also we probably still want to display IdentifierMismatch UI for the user to simplify things anyway.

Reply:
It is not 100% accurate for Advanced Shamir either: the mismatch could be in exponent, not identifier. Maybe a more precise name would be something like "ShareCollectionMismatch", as in "you are entering a share that does not belong" -- but ISTM that `IdentifierMismatch` is close enough. The UI text says as much too. So I'd keep it.